### ISSUE TITLE
Return type not inferred correctly with `__rmul__`

### ISSUE THREAD

**Issue by orip** [microsoft/pyright#956]
**Describe the bug**
pyright complains about a wrong return type, missing the result of an `__rmul__` definition.

**To Reproduce**
Analyze the sample code, receive a type error.

**Expected behavior**
Receive no type error when analyzing the code.

**Screenshots or Code**
```python
class Foo(int):
    def __mul__(self, other):
        return Foo(super().__mul__(other))
    def __rmul__(self, other):
        return self * other

X = Foo(7)

assert type(X * 10) is Foo  # __mul__
assert type(10 * X) is Foo  # __rmul__

def func(a: Foo) -> None:
    pass

# No error reported
func(X * 10)

# Marks the expression "10 * X" and says:
# Argument of type "int" cannot be assigned to parameter "a" of type "Foo" in function "func"
#  "int" is incompatible with "Foo"Pyright (reportGeneralTypeIssues)
func(10 * X)
```

**VS Code extension or command-line**
Running Pyright 1.1.63 as a VS Code extension.

**Additional context**

**Comment by erictraut** [comment]
The builtins type stub (builtins.pyi) defines the `int` class, and it contains the method `__mul__` with the following definition:
```
def __mul__(self, x: int) -> int: ...
```
For the expression `10 * X`, the LHS expression evaluates to type `int`, and the RHS expression is type `Foo` which derives from `int`. Therefore, the `__mul__` method applies, and the resulting type is `int`.

There is no type inference being used here. All of the types are known, so inference is not required. Pyright is simply applying the type rules from the builtins.pyi stub.

It appears that the actual implementation of the `__mul__` method in the built-in `int` class does not accept objects that derive from `int` as is indicated in the type stub. This is either a bug in the Python interpreter or in the typeshed `builtins.pyi` stub. From a type checking standpoint, pyright is doing the right thing.

Here's a suggested workaround:
```
_T = TypeVar("_T")
class Foo:
    def __init__(self, val: int):
        self._value = val

    def __mul__(self: _T, other: int) -> _T:
        return Foo(self._value * other)

    def __rmul__(self: _T, other: int) -> _T:
        return self * other

    def __int__(self):
        return self._value
```
